pkg/storage/read_as_of_iterator.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package storage
+
+import "github.com/cockroachdb/cockroach/pkg/util/hlc"
+
+// ReadAsOfIterator wraps an iterator that implements the SimpleMVCCIterator
```

Comment:
Before I respond to the other feedback, we should get on the same page about this. Here's my intention:
- this iterator can surface all versions of a key below a timestamp, via the `Next` function. In `TestReadAsOfIterator`, the following test: 
    ```
    // ensure next key captures at most one mvcc key per key after an asOf skip
    {input: "b3c2c1", expectedNextKey: "c2", expectedNext: "c2c1", asOf: "2"},
    ```

- ...as long as those keys are _not_ shadowed by a tombstone below the asOf timestamp
    ```
   // ensure tombstone is skipped after an AOST skip
   {input: "b3c2Xc1d1", expectedNextKey: "d1", expectedNext: "d1", asOf: "2"},
   {input: "b3c2Xc1d1", expectedNextKey: "c1d1", expectedNext: "c1d1", asOf: "1"},
    ```

- the iterator ignores _all_ tombstones above the timestamp. See tests: 
    ```
    // ensure an AOST 'next' takes precedence over a tombstone 'nextkey'
   {input: "b2Xb1c1", expectedNextKey: "c1", expectedNext: "c1", asOf: ""},
   {input: "b2Xb1c1", expectedNextKey: "b1c1", expectedNext: "b1c1", asOf: "1"},
   ```

   and in the seek test
   ```
   // Ensure AOST 'next' takes precendence over tombstone 'nextkey'
  {"a4a3Xa1b1", "a3", "a1", "1"},
  {"a4a3Xa2a1b1", "a2", "a1", "1"},
  {"a4a3Xa2a1b1", "a2", "a2", "2"},
   ```

- the iterator will surface any keys that shadow a tombstone iff those keys are below the asOf timestamp. (I now realize I shouldn't use the word 'before' in this code comment). 
   ```
  // Ensure key before delete tombstone gets read if under AOST
  {input: "b2b1Xc1", expectedNextKey: "b2c1", expectedNext: "b2c1", asOf: ""},
  {input: "b2b1Xc1", expectedNextKey: "c1", expectedNext: "c1", asOf: "1"},
  ```
  
  To obey these invariants, the logic in `SeekGE` is a bit weird. Specifically, SeekGE needs to ensure we do _not_ return a key that's shadowed by a tombstone that's below the asOf timestamp.

Reply:
I see, thanks for clarifying. I'm curious why we don't care about old versions below a tombstone, but we do if they're below a visible key -- what's the use-case here? Naïvely, I would think we'd either not care about old versions at all, or we'd care about all of them.

>To obey these invariants, the logic in SeekGE is a bit weird, but the https://github.com/cockroachdb/cockroach/pull/77281#discussion_r873765645 you propose is correct. Specifically, SeekGE needs to ensure we do not return a key that's shadowed by a tombstone that's below the asOf timestamp.

Right, makes sense. So the simplification would be this one then (which potentially avoids a lot of wasted work)?

> this can be simplified by using the lowest timestamp value of `originalKey.Timestamp` and `f.asOf` and calling `advance()` once

Reply:
>  I'm curious why we don't care about old versions below a tombstone, but we do if they're below a visible key -- what's the use-case here?

Yeah, this is an interesting point. I built this iterator to reflect the logic previously handled in `restore_processor.go`. This includes:
- Skipping to the next key if we see a tombstone below the asOf timestamp. I.e., we don't care about old versions below a tombstone.
   ```
   if len(iter.UnsafeValue()) == 0 {
			// Value is deleted.
			iter.NextKey()
			continue
		}
   ``` 
- Now, should we surface keys below a visible key? I think we should, even though the old restore processor logic didn't have any guidance on this. Currently, as soon as it finds a valid visible key, it calls `NextKey`. In other words, we _currently_ only restore at most  one version of  each key. But I could imagine a future version of Restore that requires restoring multiple versions of a key (eg https://github.com/cockroachdb/cockroach/issues/76722, which will allow a restoring cluster to finish or cancel an import from a backup cluster). 

On your other question:
> So the simplification would be this one then (which potentially avoids a lot of wasted work)?

Yes, exactly!

Reply:
>I could imagine a future version of Restore that requires restoring multiple versions of a ke

In these cases, wouldn't we want to also restore versions below tombstones? Otherwise, we'd have an incomplete history.

Anyway, I'll defer to @dt on this. I don't really have a strong opinion on these semantics as long as they cover our needs and they're consistent and well-documented.

Reply:
just spoke with @dt about this. He doesn't actually think RESTORE should restore multiple versions of the same key. (my understanding of the import issue I tagged was incorrect). So for the purposes of restore, the `readAsOfIterator` should only ever surface one valid version per key, implying `Next` and `NextKey` should have the same implementation.